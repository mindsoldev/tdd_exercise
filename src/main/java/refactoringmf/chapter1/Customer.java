package refactoringmf.chapter1;

import java.util.ArrayList;
import java.util.List;

public class Customer {

    private String name;
    private List<Rental> rentals = new ArrayList<Rental>();

    public Customer(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void addRental(Rental rental) {
        rentals.add(rental);
    }

    public String statement() {
        String result = "Rental record for " + getName() + "\n";
        for (Rental rental : rentals) {
            result += "\t" + rental.getMovie().getTitle() + "\t" + String.valueOf(rental.getCharge()) + "\n";
        }

        result += "Amount owed is " + String.valueOf(getTotalCharge()) + "\n";
        result += "You earned " + String.valueOf(getTotalFrequentRenterPoints()) + " frequent renter points";

        return result;
    }

    public String htmlStatement() {
    	String result = "<!DOCTYPE html>\n";
    	result += "<html>\n";
    	result += "<body>\n";
    	result += "\n";
    	result += "<Hl>Rentals for <EM>" + getName() + "</EM></Hl><P>\n" ;
    	for (Rental rental : rentals) {
    		result += rental.getMovie().getTitle()+ ": " + String.valueOf(rental.getCharge()) + "<BR>\n";
    	}
    	
    	result += "<P>You owed is <EM>" + String.valueOf(getTotalCharge () ) + "</EM><P>\n";
		result += "On this rental you earned <EM>" + String.valueOf(getTotalFrequentRenterPoints() ) + "</EM> frequent renter points<P>";    	

		result += "\n\n";
    	result += "</body>\n";
    	result += "</html>\n";

		return result;
    }
    
    private double getTotalCharge() {
    	double result = 0;
        for (Rental rental : rentals) {
        	result += rental.getCharge();
        }
    	return result;
    }
    
    private int getTotalFrequentRenterPoints() {
    	int result = 0;
    	for (Rental rental : rentals) {
    		result += rental.getFrequentRenterPoints();
    	}
    	return result;
    }
    
}
