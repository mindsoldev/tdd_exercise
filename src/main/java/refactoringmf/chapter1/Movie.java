package refactoringmf.chapter1;

public class Movie {

    public static final int CHILDREN = 2;
    public static final int REGULAR = 0;
    public static final int NEW_RELEASE = 1;

    private String title;
//    private int priceCode;
    private Price price;

    public Movie(String title, int priceCode) {
        this.title = title;
        setPriceCode(priceCode);
    }

    public String getTitle() {
        return title;
    }

    public int getPriceCode() {
        return price.getPriceCode();
    }

	public void setPriceCode(int priceCode) {
		switch (priceCode) {
		case REGULAR:
			price = new RegularPrice();
			break;
		case CHILDREN:
			price = new ChildrensPrice();
			break;
		case NEW_RELEASE:
			price = new NewReleasePrice();
			break;
		default:
			throw new IllegalArgumentException("Incorrect Price Code");
		}
	}

	public double getCharge(int daysRented) {
		return price.getCharge(daysRented);
	}

	int getFrequentRenterPoints(int daysRented) {
		return price.getFrequentRenterPoints(daysRented);
	}

}
