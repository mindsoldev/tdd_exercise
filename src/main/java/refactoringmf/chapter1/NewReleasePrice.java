package refactoringmf.chapter1;

public class NewReleasePrice extends Price {

	@Override
	int getPriceCode() {
		return Movie.NEW_RELEASE;
	}

	public double getCharge(int daysRented) {
		return daysRented * 3;
	}

	public int getFrequentRenterPoints(int daysRented) {
		return (daysRented >1) ? 2: 1;
	}

}
