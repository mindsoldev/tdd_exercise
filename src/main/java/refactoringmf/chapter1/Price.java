package refactoringmf.chapter1;

public abstract class Price {

	abstract int getPriceCode();

	abstract public double getCharge(int daysRented);

	int getFrequentRenterPoints(Movie movie, int daysRented) {
		return getFrequentRenterPoints(daysRented);
	}

	int getFrequentRenterPoints(int daysRented) {
		return 1;
	}
	
}
