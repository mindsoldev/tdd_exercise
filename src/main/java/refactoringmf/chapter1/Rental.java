package refactoringmf.chapter1;

public class Rental {

    private Movie movie;
    private int daysRented;

    public Rental(Movie movie, int daysRented) {
        this.movie = movie;
        this.daysRented = daysRented;
    }

    public Movie getMovie() {
        return movie;
    }

    public int getDaysRented() {
        return daysRented;
    }
    
    double getCharge() {
    	return movie.getCharge(daysRented) ;
    }
    
	int getFrequentRenterPoints() {
		return movie.getFrequentRenterPoints(daysRented);
	}

}
