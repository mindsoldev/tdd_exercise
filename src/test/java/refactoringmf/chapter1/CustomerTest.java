package refactoringmf.chapter1;


import org.junit.jupiter.api.Test;

import refactoringmf.chapter1.Customer;
import refactoringmf.chapter1.Movie;
import refactoringmf.chapter1.Rental;

import static org.assertj.core.api.Assertions.*;
import static org.approvaltests.Approvals.verify;

public class CustomerTest {

    private static final Movie THE_HULK = new Movie("The Hulk", Movie.CHILDREN);
    private static final Movie IRON_MAN = new Movie("Iron Man 4", Movie.NEW_RELEASE);
    private static final Movie SPIDER_MAN = new Movie("Spiderman", Movie.REGULAR);

    private final Customer customer = new Customer("fred");

    
    @Test
    public void basicChildrenRental() {
        customer.addRental(new Rental(THE_HULK, 2));
        assertThat(customer.statement()).isEqualTo(expectedMessageFor("The Hulk", 1.5, 1.5, 1));
    }

    @Test
    public void shouldDiscountChildrensRentals() {
        customer.addRental(new Rental(THE_HULK, 4));
        assertThat(customer.statement()).isEqualTo(expectedMessageFor("The Hulk", 3.0, 3.0, 1));
    }

    @Test
    public void basicNewReleaseRental() {
        customer.addRental(new Rental(IRON_MAN, 1));
        assertThat(customer.statement()).isEqualTo(expectedMessageFor("Iron Man 4", 3.0, 3.0, 1));
    }

    @Test
    public void shouldNotDiscountNewReleaseRentalsButBonusFrequentRenterPoints() {
        customer.addRental(new Rental(IRON_MAN, 4));
        assertThat(customer.statement()).isEqualTo(expectedMessageFor("Iron Man 4", 12.0, 12.0, 2));
    }

    @Test
    public void basicRegularRental() {
        customer.addRental(new Rental(SPIDER_MAN, 2));
        assertThat(customer.statement()).isEqualTo(expectedMessageFor("Spiderman", 2.0, 2.0, 1));
    }

    @Test
    public void shouldDiscountRegularRental() {
        customer.addRental(new Rental(SPIDER_MAN, 4));
        assertThat(customer.statement()).isEqualTo(expectedMessageFor("Spiderman", 5.0, 5.0, 1));
    }

    @Test
    public void shouldSumVariousRentals() {
        customer.addRental(new Rental(THE_HULK, 2));
        customer.addRental(new Rental(SPIDER_MAN, 1));
        customer.addRental(new Rental(IRON_MAN, 3));
        assertThat(customer.statement()).isEqualTo("Rental record for fred\n\tThe Hulk\t1.5\n\tSpiderman\t2.0\n\tIron Man 4\t9.0\nAmount owed is 12.5\nYou earned 4 frequent renter points");
    }

    private static String expectedMessageFor(String rental, double price, double total, int renterPointsEarned) {
        return "Rental record for fred\n\t" + rental + "\t" + price + "\nAmount owed is " + total + "\nYou earned " + renterPointsEarned + " frequent renter points";
    }
	
    
    @Test
	void testApprovals() throws Exception {
        StringBuilder toVerifySB = new StringBuilder();
        customer.addRental(new Rental(THE_HULK, 2));
        toVerifySB.append(customer.statement());

        verify(toVerifySB.toString());
	}

    @Test
    void testApprovalsHTML() throws Exception {
    	StringBuilder toVerifySB = new StringBuilder();
    	customer.addRental(new Rental(THE_HULK, 2));
    	toVerifySB.append(customer.htmlStatement());
    	
    	verify(toVerifySB.toString());
    }
}
