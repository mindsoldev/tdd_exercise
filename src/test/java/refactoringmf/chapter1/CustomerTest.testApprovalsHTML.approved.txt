<!DOCTYPE html>
<html>
<body>

<Hl>Rentals for <EM>fred</EM></Hl><P>
The Hulk: 1.5<BR>
<P>You owed is <EM>1.5</EM><P>
On this rental you earned <EM>1</EM> frequent renter points<P>

</body>
</html>
